packages/react-dom/src/events/DOMModernPluginEventSystem.js

```diff
@@ -305,3 +306,46 @@ export function attachElementListener(listener: ReactDOMListener): void {
 export function detachElementListener(listener: ReactDOMListener): void {
   // TODO
 }
+
+export function accumulateTwoPhaseListeners(event: ReactSyntheticEvent): void {
+  const phasedRegistrationNames = event.dispatchConfig.phasedRegistrationNames;
+  if (phasedRegistrationNames == null) {
+    return;
+  }
+  const {bubbled, captured} = phasedRegistrationNames;
+  const dispatchListeners = [];
+  const dispatchInstances = [];
+  let node = event._targetInst;
+  let hasListeners = false;
+
+  // Accumulate all instances and listeners via the target -> root path.
+  while (node !== null) {
+    // We only care for listeners that are on HostComponents (i.e. <div>)
+    if (node.tag === HostComponent) {
+      // Standard React on* listeners, i.e. onClick prop
+      const captureListener = getListener(node, captured);
+      if (captureListener != null) {
+        hasListeners = true;
+        // Capture listeners/instances should go at the start, so we
+        // unshift them to the start of the array.
+        dispatchListeners.unshift(captureListener);
+        dispatchInstances.unshift(node);
+      }
+      const bubbleListener = getListener(node, bubbled);
+      if (bubbleListener != null) {
+        hasListeners = true;
+        // Bubble listeners/instances should go at the start, so we
+        // push them to the end of the array.
+        dispatchListeners.push(bubbleListener);
+        dispatchInstances.push(node);
```

Comment:
This looks odd. You're unshifting so that `node` comes before `listener` but you're pushing in the opposite order. Mistake?

Reply:
I’m not sure I get you. I’m unshifting to insert both the node and listener at the start during capture and pushing both during bubble.

Reply:
I am not very familiar with the surrounding code, but I assume that the sequence of node -> listener -> node -> listener is significant.

But since you do this above:
```js
dispatchListeners.unshift(captureListener);
dispatchInstances.unshift(node);
```

...and this below:
```js
  dispatchListeners.push(bubbleListener);
  dispatchInstances.push(node);
```

...you will end up with a sequence of node -> listener -> ... -> listener -> node. I assume this would break code elsewhere that destructures these tuples?

Reply:
But the nodes and listeners are pushed or unshifted into separate arrays?

Reply:
Oh...gosh. 😐Yup. I was definitely reading both as the same array, even after copy-pasting. The monospace font and my own blindness got me. Sorry for the confusion.